### PR TITLE
fix: 🐛 Upgrade style-dictionary to new major version

### LINKED ISSUE #327: fix: 🤔 Tokens generation throws errors for font-weight tokens during build
# Checklist (for the issue creator)

- [x] I have verified the bug exists in the latest version of `@synergy-design-system/components` (e.g. via `pnpm install @synergy-design-system/components`).
- [x] I can provide a simple, reproducible example of the issue (e.g. via code-sandbox or sample code, ...).
- [x] I would be willing to provide a PR for this issue.

# Version

The bug occurred with the following version of the component library: @synergy-design-system/tokens@1.4.0

# Component

The issue arises when using the following component(s): -

# Environment

The issue occurred with the following operating system(s): MacOS Sonoma
The issue occurred in the following browser(s): Node.js 20 (lts)

# Description

When trying to build the tokens, there are errors on the console about tokens that cannot be found. After checking the created build artifacts, it is clear that the variables are available and are therefore also used.

Log output:

```bash

> @synergy-design-system/tokens@1.4.0 build /REDACTED/synergy/packages/tokens
> pnpm clean && node scripts/build.js


> @synergy-design-system/tokens@1.4.0 clean /REDACTED/synergy/packages/tokens
> rimraf build

Error: Reference doesn't exist: tries to reference font-weight.normal, which is not defined.
    at file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/bundled_CJS_deps/style-dictionary/lib/utils/references/resolveReferences.js:182:15
    at String.replace (<anonymous>)
    at _resolveReferences (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/bundled_CJS_deps/style-dictionary/lib/utils/references/resolveReferences.js:85:9)
    at resolveReferences (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/bundled_CJS_deps/style-dictionary/lib/utils/references/resolveReferences.js:40:10)
    at recurse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:21:32)
    at recurse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:52:13)
    at recurse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:52:13)
    at addFontStyles (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:58:5)
    at parseTokens (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/parse-tokens.js:7:28)
    at parse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/registerTransforms.js:57:28)

css
✔︎ ./dist/themes/dark.css
Error: Reference doesn't exist: tries to reference font-weight.normal, which is not defined.
    at file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/bundled_CJS_deps/style-dictionary/lib/utils/references/resolveReferences.js:182:15
    at String.replace (<anonymous>)
    at _resolveReferences (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/bundled_CJS_deps/style-dictionary/lib/utils/references/resolveReferences.js:85:9)
    at resolveReferences (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/bundled_CJS_deps/style-dictionary/lib/utils/references/resolveReferences.js:40:10)
    at recurse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:21:32)
    at recurse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:52:13)
    at recurse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:52:13)
    at addFontStyles (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/add-font-styles.js:58:5)
    at parseTokens (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/parsers/parse-tokens.js:7:28)
    at parse (file:///REDACTED/synergy/node_modules/.pnpm/@tokens-studio+sd-transforms@0.14.2_quill-delta@5.1.0_rxjs@7.8.1_tslib@2.6.2/node_modules/@tokens-studio/sd-transforms/dist/src/registerTransforms.js:57:28)

css
✔︎ ./dist/themes/light.css
✔︎ Created javascript exports
✔︎ Created typescript types
✔︎ Created SCSS exports
✔︎ Missing tokens added
```

# Acceptance criteria

- [ ] All DoD items successfully completed

### LINKED ISSUE #591: fix:  🐛  Upgrade style-dictionary to new major version
# Description

We currently use outdated versions of `style-dictionary` (3.9.2 -> 4.2.0) and `@tokens-studio/sd-transforms` (0.14.4 -> 1.2.3). There are [multiple incompatible configuration changes](https://v4.styledictionary.com/version-4/migration/) that mean we will need to spend more time with migration. It also means the token structure likely has to be changed or at least adopted (e.g. the CTI tokens will not be taken from the structure start path anymore, as documented at https://v4.styledictionary.com/version-4/migration/#cti-reliance).

# Rationale

We want to stay on the current version of style dictionary to make sure we receive the latest features. Also, the tokens-studio transforms will tie into Figma's Token-Studio much better with the lastest version.

# Acceptance criteria

- [x] upgrade core
- [x] upgrade style dictionary
- [x] upgrade matchers
- [x] upgrade external plugins
- [x] check if token structure must be updated

### PR DESCRIPTION
**Pull request by schilchSICKAG**
# Pull Request

## 📖 Description

This PR upgrades our tokens package to use the latest version `style-dictionary` and all of its plugins. It also includes a bugfix for the `js` and `scss` output targets not having the fallback tokens included (hence the change from `chore` to `fix`).

### 🎫 Issues

Closes #591 
Closes #327 

## 👩‍💻 Reviewer Notes

I have created a separate package for this to try out, so the changed files list may be rather... strange. When I was ready, I copied over the contents from the "new" package and made a final git diff.

I also made sure to allow using types by configuring typescript to allow our build scripts as linted source. This means you can write regular javascript, but still have types without a build step.

## 📑 Test Plan

First, make sure to go to main and save a copy of the `packages/tokens/dist` directory for reference.

Check out this branch (mandatory!) and install everything. Run `pnpm clean` to make sure to rebuild the package, as there are many deps updated here. After the run, you will have a new `packages/tokens/dist` directory that follows the original format and file system layout.

Check the output for visual differences (e.g. on the command line using `diff -r packages/tokens/dist PATH_TO_REFERENCE/dist`

> Please note that for the `js` and `scss` outputs, the variables for fallback figma tokens where missing in the old package. This means you **WILL** get diff messages for those packages, but only the new tokens should be displayed (e.g. starting with `SynBorderRadiusLarge`)

Also note the compilation errors are finally gone :)

## ✅ DoD

- [x] I have tested my changes manually.
- [x] I have added automatic tests for my changes (unit- and visual regression tests).
- [x] I have updated the project documentation to reflect my changes (e.g. CHANGELOG, Storybook, ...).
- [ ] ~~I have added documentation to the DaVinci migration guide (if need be)~~
- [x] I have made sure to follow the projects coding and contribution guides.
- [x] I have made sure that the bundle size has changed appropriately.
- [x] I have validated that there are no accessibility errors.
- [x] I have used design tokens instead of fix css values

### CODE REVIEW

File: packages/tokens/scripts/formats/css-vars.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-return */
+import { fileHeader, formattedVariables } from 'style-dictionary/utils';
+
+/**
+ * Adjusts color tokens to use the `color` prefix
+ * Before: syn-primary-50
+ * After: syn-color-primary-50
```
[the remark this answers is not in the document]
Comment (schilchSICKAG) [reply packages/tokens/scripts/formats/css-vars.js:10]: Fixed in latest

File: packages/tokens/scripts/transforms/add-missing-quotes.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+
+/**
+ * Add support for fallback fonts to the font-family tokens
```
[the remark this answers is not in the document]
Comment (schilchSICKAG) [reply packages/tokens/scripts/transforms/add-missing-quotes.js:6]: Fixed in latest

File: packages/tokens/scripts/transforms/adjust-shadow.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-return */
+
+/**
+ * Removes the unwanted "dropShadow" prefix from the shadow value.
+ * @type import('style-dictionary/types').ValueTransform addFallbackFonts
```
[the remark this answers is not in the document]
Comment (schilchSICKAG) [reply packages/tokens/scripts/transforms/adjust-shadow.js:7]: Fixed in latest

File: packages/tokens/scripts/transforms/use-css-calc.js [hunk 4]
```diff
@@ -36,25 +37,37 @@ const supportedPrefixes = [
 ];
 
 /**
- * Custom transform used for adding calc() around variables that need it.
- * @see https://github.com/amzn/style-dictionary/issues/820
+ * Add support for fallback fonts to the font-family tokens
+ * @type import('style-dictionary/types').ValueTransform addFallbackFonts
```
[the remark this answers is not in the document]
Comment (schilchSICKAG) [reply packages/tokens/scripts/transforms/use-css-calc.js:41]: Fixed in latest

File: packages/tokens/scripts/build.js [hunk 5]
```diff
@@ -1,104 +1,119 @@
...
+          format: 'syn/css-variable-formatter',
+          options: {
+            fileHeader: 'syn/header',
+            prefix: config.prefix,
+            theme,
           },
-        ],
-        prefix: config.prefix,
+        }],
+        prefix: 'syn-',
```
[the remark this answers is not in the document]
Comment (schilchSICKAG) [reply packages/tokens/scripts/build.js:70]: Good catch, fixed in latest

**Review by kirchsuSICKAG — APPROVED** [review]

**Pull request merged**